File: April-18/02-04-18/python/01-Monk-and-Prisoner-of-Azkaban.py
n = int(input())
ar = [int(x) for x in input().split()]
stack = []
front = []
back = []

for i in range(0, n):
    front.append(0)
    back.append(0)
    
for i in range(0, n):
    while stack and (ar[stack[-1]] <= ar[i]):
        q = stack.pop()
    if not stack:
        front[i] = -1
    else:
        front[i] = stack[-1] + 1
    stack.append(i) 

while(stack):
    q = stack.pop()

for i in range(n-1, -1 , -1):
    while stack and (ar[stack[-1]] <= ar[i]):
        q = stack.pop()
    if not stack:
        back[i] = -1
    else:
        back[i] = stack[-1] + 1
    stack.append(i) 

for i in range(0, n):
    print(front[i] + back[i] , end =' ')
print()

File: April-18/02-04-18/c++/01-Monk-and-Prisoner-of-Azkaban.cpp
#include<bits/stdc++.h>

using namespace std;
typedef long long ll;

int main()
{
    int n;
    cin >> n;
    ll ar[n];
    for(int i = 0; i < n; i++)
    {
        cin >> ar[i];
    }
    
    stack<ll> st;
    ll front[n] = {0};
    ll back[n] = {0};
    
    ll sum = 0;
    for(int i = 0; i < n; i++)
    {
        while (!st.empty() && ar[st.top()] <= ar[i])
            st.pop();
        
        front[i] = (st.empty())? (-1) : (st.top() + 1);
        st.push(i);
    }
    while(!st.empty())
        st.pop();
    for(int i = n-1 ; i >= 0; i--)
    {
        while(!st.empty() && ar[st.top()] <= ar[i])
            st.pop();
        
        back[i] = (st.empty())? (-1) : (st.top() + 1);
        st.push(i);
    }
    
    for(int i = 0; i < n; i++)
    {
        cout << front[i] + back[i] << " ";
    }
    cout << endl;
    
}
